fix(agw): swap python package json-pointer with jsonpointer

waqaraqeel opened the pull request:
## Summary

A fresh AGW VM coming up after `vagrant up magma`, `make run`, and `register_vm` is not able to checkin with the orc8r. `magtivate && checkin_cli.py` in the magma VM fails with the error:

```
pkg_resources.DistributionNotFound: The 'jsonpointer>1.13; extra == "format"' distribution was not found and is required by jsonschema
```
This is because the `jsonpointer` Python package has not been installed. Python dependencies are listed in `lte/gateway/python/setup.py` and the `install_requires` list has `json-pointer` instead of `jsonpointer`. The `json-pointer` package is different and is not one of our dependencies.

The `jsonpointer` package is indeed listed as one of our dependencies in `orc8r/gateway/python/setup.py`, but the version there is 1.13. The current `jsonschema` package we install requires `jsonpointer>=1.14`.

This PR fixes the typo/confusion in `lte/gateway/python/setup.py` and bumps the version for `jsonpointer`.

## Test Plan

Tested `checkin_cli.py` with `json-pointer` installed but `jsonpointer` uninstalled -> Failure
Tested `checkin_cli.py` with `jsonpointer` installed but `json-pointer` uninstalled -> Success!

Built and tested the `magma` package on a VM as instructed [here](https://docs.magmacore.org/docs/lte/build_install_magma_pkg_in_agw). 

Signed-off-by: Waqar Aqeel <waqeel@fb.com>

waqaraqeel committed:
fix: swap python package json-pointer with jsonpointer

Signed-off-by: Waqar Aqeel <waqeel@fb.com>

electronjoe commented:
@waqaraqeel The PR could probably use a bit more detail on reproduction methodology, as for somebody lacking context (e.g. me) I'm not sure what operation throws the error given above.

@pshelar are there any implications for 1.6 in this issue? I'm unfamiliar with the role of this setup script.

waqaraqeel commented:
> @waqaraqeel The PR could probably use a bit more detail on reproduction methodology, as for somebody lacking context (e.g. me) I'm not sure what operation throws the error given above.
> 
> @pshelar are there any implications for 1.6 in this issue? I'm unfamiliar with the role of this setup script.

Added context in the PR. I may be mistaken, but I think it has implications for fresh gateway deployments in 1.6. They won't be able to checkin with the orc8r. The `build-magma.sh` release script refers to `lte/gateway/python/setup.py` to build the magma package and the v1.6.0 tag has the incorrect dependency there.

pshelar commented:
It seems to be working on 1.6 agw with json-pointer, am I missing something?
```
root@phy-u3:~# dpkg --list |grep magma
ii  magma                                1.6.0-1625592603-1f26ba81         amd64        Magma Access Gateway
ii  magma-cpp-redis                      4.3.1.1-2                         amd64        Redis C++ client library
ii  magma-libfluid                       0.1.0.6-1                         amd64        Libfluid Openflow Controller
ii  magma-libtacopie                     3.2.0.1-1                         amd64        TacoPie C++ library
ii  magma-sctpd                          1.6.0-1625592603-1f26ba81         amd64        Magma SCTPD
root@phy-u3:~# checkin_cli.py
1. -- Testing TCP connection to controller.ens.fbmagma.ninja:443 --
2. -- Testing Certificate --
3. -- Testing SSL --
4. -- Creating direct cloud checkin --
5. -- Creating proxy cloud checkin --

Success!

root@phy-u3:~# dpkg --list |grep json |grep pointer
ii  python3-json-pointer                 2.0-0ubuntu1                      all          resolve JSON pointers - Python 3.x
```

themarwhal commented:
Hey @pshelar ! Just curious, do we still need the debian python dependencies? (In setup.py files.) Should we clean that up?

waqaraqeel commented:
> It seems to be working on 1.6 agw with json-pointer, am I missing something?
> 
> ```
> root@phy-u3:~# dpkg --list |grep magma
> ii  magma                                1.6.0-1625592603-1f26ba81         amd64        Magma Access Gateway
> ii  magma-cpp-redis                      4.3.1.1-2                         amd64        Redis C++ client library
> ii  magma-libfluid                       0.1.0.6-1                         amd64        Libfluid Openflow Controller
> ii  magma-libtacopie                     3.2.0.1-1                         amd64        TacoPie C++ library
> ii  magma-sctpd                          1.6.0-1625592603-1f26ba81         amd64        Magma SCTPD
> root@phy-u3:~# checkin_cli.py
> 1. -- Testing TCP connection to controller.ens.fbmagma.ninja:443 --
> 2. -- Testing Certificate --
> 3. -- Testing SSL --
> 4. -- Creating direct cloud checkin --
> 5. -- Creating proxy cloud checkin --
> 
> Success!
> 
> root@phy-u3:~# dpkg --list |grep json |grep pointer
> ii  python3-json-pointer                 2.0-0ubuntu1                      all          resolve JSON pointers - Python 3.x
> ```

So this is a Debian/Ubuntu apt package from the repositories. I was talking about a pip package. The apt package is not installed on my vagrant box though...

pshelar commented:
> Hey @pshelar ! Just curious, do we still need the debian python dependencies? (In setup.py files.) Should we clean that up?

yes, lets clean it up. it will vastly improve readability.

pshelar commented:
> > It seems to be working on 1.6 agw with json-pointer, am I missing something?
> > ```
> > root@phy-u3:~# dpkg --list |grep magma
> > ii  magma                                1.6.0-1625592603-1f26ba81         amd64        Magma Access Gateway
> > ii  magma-cpp-redis                      4.3.1.1-2                         amd64        Redis C++ client library
> > ii  magma-libfluid                       0.1.0.6-1                         amd64        Libfluid Openflow Controller
> > ii  magma-libtacopie                     3.2.0.1-1                         amd64        TacoPie C++ library
> > ii  magma-sctpd                          1.6.0-1625592603-1f26ba81         amd64        Magma SCTPD
> > root@phy-u3:~# checkin_cli.py
> > 1. -- Testing TCP connection to controller.ens.fbmagma.ninja:443 --
> > 2. -- Testing Certificate --
> > 3. -- Testing SSL --
> > 4. -- Creating direct cloud checkin --
> > 5. -- Creating proxy cloud checkin --
> > 
> > Success!
> > 
> > root@phy-u3:~# dpkg --list |grep json |grep pointer
> > ii  python3-json-pointer                 2.0-0ubuntu1                      all          resolve JSON pointers - Python 3.x
> > ```
> 
> So this is a Debian/Ubuntu apt package from the repositories. I was talking about a pip package. The apt package is not installed on my vagrant box though...

ok, So its not 1.6 issue then.
Lets keep vagrant image and deployed AGW as close as possible. Can you compare which package it better fit for AGW. We can make sure we have that package across all AGW setups.

waqaraqeel commented:
@pshelar This is a funny coincidence. The confusion between `json-pointer` and `jsonpointer` did not become an issue because there happens to be an apt package by the name of `python3-json-pointer` which actually provides the PyPi package `jsonpointer` and not `json-pointer`. In pydep, when [`gen_sys_package_name`](https://github.com/magma/magma/blob/master/lte/gateway/release/pydep#L148) gets called with `py_pkgname="json-pointer"`, it generates an apt package name that exists and happens to be the right one.

I still think we should make the correction in `setup.py` though as this PR does. In general, I don't know wether the assumption that a `python3-{dep}` apt package, if it exists, will provide the right PyPi package `{dep}` is valid. We should think about removing this assumption. WDYT?

cc: @electronjoe @themarwhal.

pshelar commented:
> @pshelar This is a funny coincidence. The confusion between `json-pointer` and `jsonpointer` did not become an issue because there happens to be an apt package by the name of `python3-json-pointer` which actually provides the PyPi package `jsonpointer` and not `json-pointer`. In pydep, when [`gen_sys_package_name`](https://github.com/magma/magma/blob/master/lte/gateway/release/pydep#L148) gets called with `py_pkgname="json-pointer"`, it generates an apt package name that exists and happens to be the right one.
> 
> I still think we should make the correction in `setup.py` though as this PR does. In general, I don't know wether the assumption that a `python3-{dep}` apt package, if it exists, will provide the right PyPi package `{dep}` is valid. We should think about removing this assumption. WDYT?
> 
> cc: @electronjoe @themarwhal.

we have seen such discrepancy: https://github.com/magma/magma/blob/ce2d760775dc72c57399d2d229cc6a0ceb24e7f8/lte/gateway/release/pydep#L71

I am fine with the change, but before landing the change lets validate it by building the magma package and deploying in on agw.

waqaraqeel committed:
refactor: include generated magma.lockfile
                                          Signed-off-by: Waqar Aqeel <waqeel@fb.com>

Signed-off-by: Waqar Aqeel <waqeel@fb.com>

waqaraqeel commented:
> we have seen such discrepancy:
> 
> https://github.com/magma/magma/blob/ce2d760775dc72c57399d2d229cc6a0ceb24e7f8/lte/gateway/release/pydep#L71
> 
> I am fine with the change, but before landing the change lets validate it by building the magma package and deploying in on agw.

Thanks for the clarification @pshelar! I have added a mapping in `PYPI_TO_DEB` and built and tested the `magma` package on gateway VM. Added to PR test plan.

pshelar reviewed (approved):
LGTM

themarwhal reviewed (approved):
looks good! Thanks for making the changing + validating all the steps!
